rmm/monitor/src/rmm/granule.rs

```diff
@@ -41,209 +41,146 @@ impl Default for GranuleState {
+        unsafe {
+            core::ptr::write_bytes(buf as *mut usize, 0x0, GRANULE_SIZE / 8);
         }
     }
 }
 
-// Define a function find_granule which returns a Granule instance from the GRANULE array using the given address and expected state.
-pub fn find_granule(addr: usize, expected_state: GranuleState) -> Option<&'static mut Granule> {
-    let idx = granule_addr_to_idx(addr);
-    if idx >= RMM_MAX_GRANULES {
-        warn!("check the granule index {}", idx);
-        return None;
+// Implement set_granule for Granule state control and check the valid.
+pub fn set_granule(addr: usize, state: GranuleState, mm: PageMap) {
+    let granule = find_granule(addr);
+    let prev_state = granule.get_state();
+    match state {
+        GranuleState::Delegated => {
+            if prev_state != GranuleState::Undelegated
+                && prev_state != GranuleState::Delegated
+                && prev_state != GranuleState::RTT
+            {
+                granule.zeroize();
+                mm.unmap(addr);
+            }
+            granule.set_state(state);
+        }
+        GranuleState::RTT => granule.set_state(state),
+        GranuleState::Undelegated => unsafe {
+            // TODO: after RIPAS implement, it will be changed to panic
+            if prev_state != GranuleState::Delegated {
+                warn!(
+                    "granule[{:X}]? expected:Delegated->found:{:?}",
+                    addr, prev_state
+                );
+            }
+            granule.set_state(state);
+            GRANULE.remove(&granule.addr());
+        },
+        _ => {
+            // TODO: after RIPAS implement, it will be changed to panic
+            if prev_state != GranuleState::Delegated {
+                warn!(
+                    "granule[{:X}]? expected:Delegated->found:{:?}",
+                    addr, prev_state
+                );
+            }
+            granule.set_state(state);
+            mm.map(addr, true);
+        }
     }
+}
 
+// Define a function find_granule which returns a Granule instance from the GRANULE array using the given address and expected state.
+fn find_granule(addr: usize) -> &'static mut Granule {
+    check_validate_addr(addr);
     unsafe {
-        match GRANULE.get_mut(&idx) {
-            Some(g) => {
-                if expected_state != g.get_state() {
-                    info!(
-                        "check the {:X} granule state {:?}<-{:?}",
-                        addr,
-                        g.get_state(),
-                        expected_state
-                    );
-                }
-                Some(g)
-            }
+        match GRANULE.get_mut(&addr) {
+            Some(g) => g,
             None => {
                 let new = Granule {
-                    state: expected_state,
+                    state: GranuleState::Undelegated,
                     // refcount: 0,
-                    idx: idx,
+                    addr: addr,
```

Comment:
We can init members [shorthand](https://rust-lang.github.io/rfcs/1682-field-init-shorthand.html) way. And it is preferred style :)

```suggestion
                    addr,
```

Reply:
applied it